File: src/generator_oak.cpp
#include <iostream>
#include <string>
#include <sys/stat.h>
#include "depthai/depthai.hpp"
#include <opencv2/opencv.hpp>

int main() {
    std::string save_dir = "../data/image/chessboard_10_7_paper_st_1";
    mkdir(save_dir.c_str(), 0777);

    dai::Pipeline pipeline;

    auto monoLeft = pipeline.create<dai::node::MonoCamera>();
    auto monoRight = pipeline.create<dai::node::MonoCamera>();

    monoLeft->setResolution(dai::MonoCameraProperties::SensorResolution::THE_720_P);
    monoLeft->setBoardSocket(dai::CameraBoardSocket::CAM_B);
    monoRight->setResolution(dai::MonoCameraProperties::SensorResolution::THE_720_P);
    monoRight->setBoardSocket(dai::CameraBoardSocket::CAM_C);

    auto xoutLeft = pipeline.create<dai::node::XLinkOut>();
    auto xoutRight = pipeline.create<dai::node::XLinkOut>();
    xoutLeft->setStreamName("left");
    xoutRight->setStreamName("right");

    monoLeft->out.link(xoutLeft->input);
    monoRight->out.link(xoutRight->input);

    dai::Device device(pipeline);

    auto qLeft = device.getOutputQueue("left", 4, false);
    auto qRight = device.getOutputQueue("right", 4, false);

    int count = 1;

    while (true) {
        auto inLeft = qLeft->get<dai::ImgFrame>();
        auto inRight = qRight->get<dai::ImgFrame>();

        cv::Mat frameLeft(inLeft->getHeight(), inLeft->getWidth(), CV_8UC1, inLeft->getData().data());
        cv::Mat frameRight(inRight->getHeight(), inRight->getWidth(), CV_8UC1, inRight->getData().data());

        cv::Mat combined;
        cv::hconcat(frameLeft, frameRight, combined);
        cv::imshow("Press SPACE to capture | 'q' to quit", combined);

        int key = cv::waitKey(1);
        if (key == 32) {
            cv::imwrite(save_dir + "/left" + std::to_string(count) + ".png", frameLeft);
            cv::imwrite(save_dir + "/right" + std::to_string(count) + ".png", frameRight);
            std::cout << "Saved pair " << count << std::endl;
            count++;
        } else if (key == 'q') {
            break;
        }
    }

    cv::destroyAllWindows();
    return 0;
}

File: python/generator_oak.py
import depthai as dai
import cv2
import os

save_dir = "../data/image/chessboard_10_7_paper_st_1"
os.makedirs(save_dir, exist_ok=True)

pipeline = dai.Pipeline()

monoLeft = pipeline.create(dai.node.MonoCamera)
monoRight = pipeline.create(dai.node.MonoCamera)

monoLeft.setResolution(dai.MonoCameraProperties.SensorResolution.THE_720_P)
monoLeft.setBoardSocket(dai.CameraBoardSocket.LEFT)
monoRight.setResolution(dai.MonoCameraProperties.SensorResolution.THE_720_P)
monoRight.setBoardSocket(dai.CameraBoardSocket.RIGHT)

xoutLeft = pipeline.create(dai.node.XLinkOut)
xoutRight = pipeline.create(dai.node.XLinkOut)
xoutLeft.setStreamName("left")
xoutRight.setStreamName("right")

monoLeft.out.link(xoutLeft.input)
monoRight.out.link(xoutRight.input)

count = 1

with dai.Device(pipeline) as device:
    qLeft = device.getOutputQueue(name="left", maxSize=4, blocking=False)
    qRight = device.getOutputQueue(name="right", maxSize=4, blocking=False)

    while True:
        inLeft = qLeft.get()
        inRight = qRight.get()

        frameLeft = inLeft.getCvFrame()
        frameRight = inRight.getCvFrame()

        combined = cv2.hconcat([frameLeft, frameRight])
        cv2.imshow("Press SPACE to capture | 'q' to quit", cv2.resize(combined, (800,400)))

        key = cv2.waitKey(1)
        if key == 32: # space
            cv2.imwrite(f"{save_dir}/left{count}.png", frameLeft)
            cv2.imwrite(f"{save_dir}/right{count}.png", frameRight)
            print(f"Saved pair {count}")
            count += 1
        elif key == ord('q'):
            break

cv2.destroyAllWindows()